### PR TITLE
Generate P/Invoke transitions for CoreRT.

### PR DESCRIPTION
**Pull request by pgavlin**
This change adds support for CoreRT-style P/Invoke transitions
to RyuJIT. Instead of the usual inlined transition code, these
transitions are made up of calls to two new JIT helpers:

    PInvokeTransitionFrame frame; // opaque local
    CORINFO_HELP_INIT_PINVOKE_FRAME(&frame);
    ...
    CORINFO_HELP_JIT_PINVOKE_BEGIN(&frame);
    target(...);
    CORINFO_HELP_JIT_PINVOKE_END(&frame);
    ...

The preemptive mode constraints apply between calls to
JIT_PINVOKE_BEGIN and JIT_PINVOKE_END: no managed references
may be live only in registers and managed references may not
be manipulated.

Note that this is a rebased version of the code from https://github.com/dotnet/coreclr/pull/2817.

### CODE REVIEW

**Comment by pgavlin** [comment]
@jkotas @dotnet/jit-contrib PTAL

**Comment by pgavlin** [comment]
@BruceForstall PTAL

File: src/inc/corjit.h [hunk 1]
```diff
@@ -123,16 +123,14 @@ enum CorJitFlag
...
 
-#ifdef MDIL
     // Safe to overlap with CORJIT_FLG_MAKEFINALCODE below. Not used by the JIT, used internally by NGen only.
     CORJIT_FLG_MINIMAL_MDIL        = 0x00008000, // Generate MDIL code suitable for use to bind other assemblies.
 
     // Safe to overlap with CORJIT_FLG_READYTORUN below. Not used by the JIT, used internally by NGen only.
     CORJIT_FLG_NO_MDIL             = 0x00010000, // Generate an MDIL section but no code or CTL. Not used by the JIT, used internally by NGen only.
+#else // MDIL
+    CORJIT_FLG_CFI_UNWIND          = 0x00004000, // Emit CFI unwind info
 #endif // MDIL
```
Comment (BruceForstall) [inline src/inc/corjit.h]: Why did you change this? It was intentional as it was, to keep identical bit patterns together.

File: src/jit/codegencommon.cpp [hunk 2]
```diff
@@ -8518,13 +8520,23 @@ void                CodeGen::genFnProlog()
...
     {
-        excludeMask |= (RBM_PINVOKE_FRAME | RBM_PINVOKE_TCB | RBM_PINVOKE_SCRATCH);
+        excludeMask |= RBM_PINVOKE_FRAME;
 
-        // We also must exclude the register used by compLvFrameListRoot when it is enregistered
-        //
-        LclVarDsc *  varDsc = &compiler->lvaTable[compiler->info.compLvFrameListRoot];
-        if (varDsc->lvRegister)
-            excludeMask |= genRegMask(varDsc->lvRegNum);
+        assert(!compiler->opts.ShouldUsePInvokeHelpers() || compiler->info.compLvFrameListRoot == BAD_VAR_NUM);
```
Comment (BruceForstall) [inline src/jit/codegencommon.cpp]: CC: fully parenthesize sub-expressions

Comment (BruceForstall) [inline src/jit/codegencommon.cpp]: There's an ordering problem here, because you already found and used a varDsc based on compLvFrameListRoot, which might be illegal if that is BAD_VAR_NUM.

Comment (pgavlin) [inline src/jit/codegencommon.cpp]: Okay.

FWIW, I cannot find this guidance in [the coding conventions](https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-jit-coding-conventions.md).

Comment (BruceForstall) [inline src/jit/codegencommon.cpp]: Maybe there should be a comment that with ShouldUsePInvokeHelpers() true, that the PCB and SCRATCH registers are not used or modified, and can be used as an initReg?

File: src/jit/compiler.h [hunk 3]
```diff
@@ -7707,6 +7707,10 @@ public :
         inline bool         IsReadyToRun() { return false; }
 #endif
 
+        // true if we should use the PINVOKE_{BEGIN,END} helpers instead of generating
+        // PInvoke transitions inline (e.g. when targeting CoreRT).
+        inline bool         ShouldUsePInvokeHelpers() { return (jitFlags->corJitFlags2 & CORJIT_FLG2_USE_PINVOKE_HELPERS) != 0; }
```
Comment (BruceForstall) [inline src/jit/compiler.h]: Please put the {, body, and }, on separate lines from the declaration.

File: src/jit/gentree.h [hunk 4]
```diff
@@ -2107,7 +2107,7 @@ struct GenTreeLclVar: public GenTreeLclVarCommon
                             DEBUG_ARG(largeNode)),
             gtLclILoffs(ilOffs)
             {
-                assert(OperIsLocal(oper));
+                assert(OperIsLocal(oper) || OperIsLocalAddr(oper));
```
Comment (BruceForstall) [inline src/jit/gentree.h]: This seems like a fundamental change: GT_LCL_var can now take a local addr? why?

File: src/inc/corjit.h (hunk 1 above)
Comment (pgavlin) [inline src/inc/corjit.h]: > Why did you change this?
This was part of the earlier change (which had `FLG_USE_PINVOKE_HELPERS` instead of `FLG_CFI_UNWIND`); I kept it when I rebased.

>  It was intentional as it was, to keep identical bit patterns together.

If we believe that keeping identical bit patterns together is important, then we should apply that to the rest of the file (e.g. the definitions for `CORJIT_FLG_MAKEFINALCODE` and `CORJIT_FLG_READYTORUN` should be next to `CORJIT_FLG_MINIMAL_MDIL` and `CORJIT_FLG_NO_MDIL`, respectively). I'm fine with either keeping all of the bits that are affected by an particular symbol together or keeping the bits themselves together. I have a minor preference for the former since it makes it easier to understand what changes when a given symbol is defined. Note that we already use that style for processor-specific preprocessor symbols.

File: src/jit/gentree.h (hunk 4 above)
Comment (pgavlin) [inline src/jit/gentree.h]: GT_LCL_var can already be a local addr, just not at the time it is constructed (i.e. it is legal to create an lcl_var node and then change its type to a lcl_addr).

File: src/jit/compiler.h (hunk 3 above)
Comment (pgavlin) [inline src/jit/compiler.h]: The method as written is consistent with the various accessors immediately above it.

Also, I cannot find any guidance re: braces for methods in [the CC](https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-jit-coding-conventions.md).

File: src/jit/regalloc.cpp [hunk 5]
```diff
@@ -6438,15 +6438,21 @@ void               Compiler::rpPredictRegUse()
...
-                // We would prefer to have this be enregister in the PINVOKE_TCB register
-                pinvokeVarDsc->addPrefReg(RBM_PINVOKE_TCB, this);
+                if (pinvokeVarDsc->lvTracked)
+                {
+                    rpRecordRegIntf(RBM_CALLEE_TRASH, VarSetOps::MakeSingleton(this, pinvokeVarDsc->lvVarIndex)
+                                    DEBUGARG("compLvFrameListRoot"));
+
+                    // We would prefer to have this be enregister in the PINVOKE_TCB register
+                    pinvokeVarDsc->addPrefReg(RBM_PINVOKE_TCB, this);
+                }
```
Comment (BruceForstall) [inline src/jit/regalloc.cpp]: this function is not used in RyuJIT, so these changes shouldn't be necessary.

**Comment by BruceForstall** [comment]
It would be nice to see a comment somewhere in the code (maybe by genPInvokeProlog(?)) that describes what the p/invoke function characteristics are (i.e., what the function looks like) with the new model, with examples.

File: src/jit/regalloc.cpp (hunk 5 above)
Comment (pgavlin) [inline src/jit/regalloc.cpp]: Ah, cool. I'll revert them.

**Comment by pgavlin** [comment]
> i.e., what the function looks like

What characteristics are you looking for? Calling convention, signature, etc.?

**Comment by pgavlin** [comment]
Or just the shape of the generated code?

File: src/jit/codegencommon.cpp (hunk 2 above)
Comment (BruceForstall) [inline src/jit/codegencommon.cpp]: https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-jit-coding-conventions.md#111-logical-and-arithmetic-expressions

Comment (pgavlin) [inline src/jit/codegencommon.cpp]: Perfect, thanks!

File: src/jit/compiler.h (hunk 3 above)
Comment (BruceForstall) [inline src/jit/compiler.h]: apparently not explicitly stated. but all examples use this form, e.g.: https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-jit-coding-conventions.md#941-example

**Comment by BruceForstall** [comment]
all of the above

File: src/jit/regalloc.cpp (hunk 5 above)
Comment (jkotas) [inline src/jit/regalloc.cpp]: It would be nice to at least leave assert here - if somebody tries to use the legacy JIT for CoreRT.

Comment (pgavlin) [inline src/jit/regalloc.cpp]: Done.

**Comment by pgavlin** [comment]
Feedback has been addressed; PTAL.

File: src/jit/compiler.h (hunk 3 above)
Comment (pgavlin) [inline src/jit/compiler.h]: I'd prefer to remain consistent with the other accessors in the case. If we believe that it is important that these methods have braces on the next line, we should change them together.

**Comment by BruceForstall** [comment]
LGTM

**Comment by pgavlin** [comment]
@dotnet-bot test this please
